peachjam/models/core_document.py

```diff
@@ -206,40 +207,73 @@ def update_languages(self):
 
     def update_extracted_citations(self):
         """Update the current work's ExtractedCitations."""
-        target_works = Work.objects.filter(
-            frbr_uri__in=self.fetch_cited_works_frbr_uris()
-        )
-
         # delete existing extracted citations
         ExtractedCitation.objects.filter(citing_work=self).delete()
 
-        for target_work in target_works:
-            ExtractedCitation.objects.get_or_create(
-                citing_work=self, target_work=target_work
+        for work_frbr_uri, treatments in self.fetch_cited_works_frbr_uris().items():
```

Comment:
Rather do the lookup of work objects in bulk:

```python
cited_works = self.fetch...()
works = {w.frbr_uri: w for w in Work.objects.filter(frbr_uri__in=cited_works.keys())}
for frbr_uri, treatments in cited_works.items():
  work = works.get(frbr_uri)
  if work:
    ExtractedCitation...
```

Reply:
updated